### ISSUE TITLE
Understanding OE Implementation of C++ Standard Library

### ISSUE THREAD

**Issue by jonahrosenblum** [openenclave/openenclave#4777]
**Summary:** I do not understand what makes something from the STL easier or harder to support in OE. I found that OE's implementation of std::exp() to be very slow in a trusted region compared to an untrusted region which is unintuitive to me.

---

Hello. I recently observed something when using OE's implementation of the STL Consider the following code:

`#include <cmath>`
`double x = 3.14`
`std::exp(x)`

I have confirmed that this code is compiled differently in trusted versus untrusted regions.

In untrusted regions, this is handled by calling the GNU lib implementation of exp

`4132b0:	ff 25 a2 5e 11 00    	jmp    *0x115ea2(%rip)        # 529158 <exp@GLIBC_2.29>`

However, in trusted regions, it seems like OE has its own exp function that is invoked

`0000000000181d40 <exp>:`
`... # rest of function`

I understand that OE has reimplemented large portions of the STL to work inside an enclave. However, I assumed that this would mostly be for things like file I/O, networking, dynamic memory allocations etc. that all require support from the OS.

The only reason I discovered this is that I have found OE's implementation of std::exp to be **notably slower** than the STL's, leading me to believe that GNU lib's implementation is not what is being used. I would love to learn more about how the STL was implemented for OE, and what challenges are present when doing this conversion because it seems like an interesting problem. I read [this table ](https://github.com/openenclave/openenclave/blob/5b3108c8cda89f48ef061d2c876f02fb89da38d4/docs/LibcxxSupport.md) and was slightly confused why things I assumed would be more difficult to support (like iostreams, atomics, etc.) are fully implemented but cmath fails to provide fully accurate implementations of functions like fmal() and tgamma().

Edit: Am I correct in thinking that [this](https://github.com/openenclave/openenclave-musl/blob/85e0e3519655220688e757b9d5bfd314923548bd/src/math/exp.c) is the exp implementation used inside the enclave?

**Comment by anakrish** [comment]
OE uses [MUSL](https://musl.libc.org/) as its C library, due to its simplicity as well as license. Similarly OE uses specific libraries for C++, exception handling, compiler runtime etc.

> I understand that OE has reimplemented large portions of the STL to work inside an enclave
> ..
>  I would love to learn more about how the STL was implemented for OE, and what challenges are present when doing this conversion because it seems like an interesting problem.

For the C++ standard library, we have mostly leveraged existing libraries with minimal changes. The [3rdparty](https://github.com/openenclave/openenclave/blob/master/3rdparty) folder contains the various libraries we use.

> OE's implementation of std::exp to be notably slower than the STL's

One experiment to try out would be to run your benchmarking program (or specifically measure exp function) within an Alpine Linux container. Alpine Linux also uses MUSL as its C library instead of GNU libc. The performance you see numbers should match the numbers you see within the enclave.  If they don't match, we may need to look at whether something else is making it slow within enclave context.

> Am I correct in thinking that [this](https://github.com/openenclave/openenclave-musl/blob/85e0e3519655220688e757b9d5bfd314923548bd/src/math/exp.c) is the exp implementation used inside the enclave?

Yes. That would the be implementation. Typically math functions have a C implementation or an assembly implementation. The assembly implementation will be used if available, for performance.

**Comment by jonahrosenblum** [comment]
@anakrish Thanks for your insight. I am working on creating microbenchmarks to make testing this performance discrepancy easier. I will look into testing in an Alpine Linux environment.

I have also found a few other things that I believe may be performance bugs in Open Enclave related to pointer dereferencing, class member variables, etc. I will create microbenchmarks related to these as well so that I can share my findings.

**Comment by jonahrosenblum** [comment]
Later today I will confirm the `exp` performance on Alpine Linux, but I actually have another observation/question.

I stole this line from an OE example at some point:

`CXXFLAGS=$(shell pkg-config oeenclave-$(CXX_COMPILER) --cflags)`

Which works well for generating flags to include all the proper OE libraries and whatnot. I also noticed that it renders this flag: `-mllvm -x86-speculative-load-hardening`. Based on the name, I suspect this is a Spectre mitigation to prevent speculation on loads by adding fences.

I noticed that this flag causes large overheads (>30%) for some applications and was wondering if OE is aware of this and wondered if this is recommended or required given that (I assume?) modern SGX-enabled CPUs would likely have Spectre mitigations from Intel.
